### DISCUSSION TITLE
Code to show what roles a user has if assigned to more than 1 teams using teams permissions

### DISCUSSION THREAD

**Discussion by gzai** [spatie/laravel-permission#2103] [Q&A]
is there a short code to show what roles a user has if assigned to more than 1 teams using teams permissions?
Need to show a user has any roles in CMS users for superadmin view or when a user choose by himself to select a team after logging in.

because I made the code it myself, will be deleted if update vendor. 
if there is no code like ``setGlobalShowAllPermissions``, all roles for one user will not appear if attached to a team.
must set for ``setPermissionsTeamId`` to show a user has any roles in that team, but it's not possible to display all user roles in CMS user for superadmin view and others.

controller
```
class DashboardFrontend extends Component
{
    public function render()
    {
        setGlobalShowAllPermissions(1);
    }
}
```

Spatie\Permission\Traits\HasRoles.php 
```
    /**
     * A model may have multiple roles.
     */
    public function roles(): BelongsToMany
    {
        $relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            PermissionRegistrar::$pivotRole
        );

        if (! PermissionRegistrar::$teams) {
            return $relation;
        }
        
        /* additional */
        if (app(PermissionRegistrar::class)->getGlobalShowAllPermissions()) {
            return $relation;
        }

        return $relation->wherePivot(PermissionRegistrar::$teamsKey, getPermissionsTeamId())
            ->where(function ($q) {
                $teamField = config('permission.table_names.roles').'.'.PermissionRegistrar::$teamsKey;
                $q->whereNull($teamField)->orWhere($teamField, getPermissionsTeamId());
            });
    }
```
Spatie\Permission\helpers.php
```
if (! function_exists('setGlobalShowAllPermissions')) {
    /**
     * @return boolean
     */
    function setGlobalShowAllPermissions($status)
    {
        app(\Spatie\Permission\PermissionRegistrar::class)->setGlobalShowAllPermissions($status);
    }
}

if (! function_exists('getGlobalShowAllPermissions')) {
    /**
     * @return boolean
     */
    function getGlobalShowAllPermissions()
    {
        return app(\Spatie\Permission\PermissionRegistrar::class)->getGlobalShowAllPermissions();
    }
}
```
Spatie\Permission\PermissionRegistrar.php
```
class PermissionRegistrar
{
    ...
    /** @var boolean */
    public $globalShowAllPermissions = false;
    ...

    /**
     * Set the global show all permissions
     *
     * @param boolean
     */
    public function setGlobalShowAllPermissions($status)
    {
        $this->globalShowAllPermissions = $status;
    }

    /**
     *
     * @return boolean
     */
    public function getGlobalShowAllPermissions()
    {
        return $this->globalShowAllPermissions;
    }
}
```

**Comment by erikn69** [comment]
Do the same idea of this https://github.com/spatie/laravel-permission/pull/1948

**Reply by gzai** [reply]
maybe a little different idea.

teams : 
| id | name |
| ------------- | ------------- |
| 1 | Team A |
| 2 | Team B |

roles : 
| id | team_id | name | guard_name |
| ------------- | ------------- | ------------- | ------------- |
| 1 | Null | root | web |
| 2 | 1 | owner | web |
| 3 | 1 | member | web |
| 4 | 2 | owner | web |
| 5 | 2 | member | web |

model_has_roles : 
| role_id | model_type | model_id | team_id |
| ------------- | ------------- | ------------- | ------------- |
| 1 | App\Models\User | 1 | Null |
| 2 | App\Models\User | 2 | 1 |
| 3 | App\Models\User | 3 | 1 |
| 5 | App\Models\User | 3 | 2 |

On CMS Users for superadmin view without my code ``setGlobalShowAllPermissions(1)``, it will display empty for user roles if there is one or more teams attached.

Controller : 
```php
$users = User::paginate( 10 );
```
View : 
```blade
@if( $users->count() > 0 )
    @foreach( $users as $key => $data )
        <tr>
            <td>{{ $data->name }}</td>
            <td>{{ $data->email }}</td>
            <td>
            @if( $data->roles->count() > 0 )
            @foreach( $data->roles as $key1 => $data1 )
                {{ $data1->name }}@if( $data1->team_id != '' ) ( {{ $data1->team->name }} )@endif
            @endforeach
            @endif
            </td>
        </tr>
    @endforeach
@endif
```
Result : 
| Name | Email | Role |
| ------------- | ------------- | ------------- |
| superadmin | superadmin@email.com | root |
| userA | userA@email.com |  |
| userB | userB@email.com |  |

On CMS Users for superadmin view with my code ``setGlobalShowAllPermissions(1)``, it will display user roles if there is one or more teams.

Controller : 
```php
setGlobalShowAllPermissions(1);

$users = User::paginate( 10 );
```
Result : 
| Name | Email | Role |
| ------------- | ------------- | ------------- |
| superadmin | superadmin@email.com | root |
| userA | userA@email.com | owner ( Team A ) |
| userB | userB@email.com | member ( Team A ), member ( Team B ) |

and for user after logging in when not already attached with ``setPermissionsTeamId($team_id)``

Controller : 
```php
// login as user B

$teams = Teams::leftjoin('team_users', 'teams.users.team_id', '=', 'teams.id')
                ->where('team_users.user_id', Auth::id() )
                ->get();
```
View : 
```blade
@if( $teams->count() > 0 )
    @foreach( $teams as $key => $data )
        <div class="teams">
            <h2>{{ $data->name }}</h2>
            <div class="detail>
            @can('team detail')
                <a href="#to-detail-link">Detail</a>
            @endcan 
            </div>
        </div>
    @endforeach
@endif
```
Result : 
| Team A | Team B |
| ------------- | ------------- |
| detail link not shown | detail link not shown |

Controller : 
```php
// login as user B
setGlobalShowAllPermissions(1);

$teams = Teams::leftjoin('team_users', 'teams.users.team_id', '=', 'teams.id')
                ->where('team_users.user_id', Auth::id() )
                ->get();
```
Result : 
| Team A | Team B |
| ------------- | ------------- |
| detail link shown | detail link shown |
